src/game/client/gui/guicallbacks/controlbarcallback.cpp

```diff
@@ -12,18 +12,360 @@
+        static const NameKeyType beaconDeleteButtonID =
+            g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonDeleteBeacon");
+        static const NameKeyType beaconClearTextButtonID =
+            g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonClearBeaconText");
+        static const NameKeyType beaconGeneralButtonID = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonGeneral");
+        static const NameKeyType buttonLargeID = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonLarge");
+        static const NameKeyType buttonOptions = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonOptions");
+        static const NameKeyType buttonIdleWorker = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonIdleWorker");
+        int id = window2->Win_Get_Window_Id();
+
+        if (id == buttonCommunicator) {
+            Toggle_Diplomacy(false);
+        } else if (id == beaconPlacementButtonID && g_theGameLogic->Is_In_Multiplayer_Game()
+            && g_thePlayerList->Get_Local_Player()->Is_Player_Active()) {
+            g_theInGameUI->Set_GUI_Command(g_theControlBar->Find_Command_Button("Command_PlaceBeacon"));
+        } else if (id == beaconDeleteButtonID && g_theGameLogic->Is_In_Multiplayer_Game()) {
+            g_theMessageStream->Append_Message(GameMessage::MSG_REMOVE_BEACON);
+        } else if (id == beaconClearTextButtonID && g_theGameLogic->Is_In_Multiplayer_Game()) {
+            static const NameKeyType textID = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:EditBeaconText");
+            GameWindow *text_window = g_theWindowManager->Win_Get_Window_From_Id(nullptr, textID);
+
+            if (text_window != nullptr) {
+                Gadget_Text_Entry_Set_Text(text_window, Utf16String::s_emptyString);
+            }
+        } else if (id == beaconGeneralButtonID) {
+            Hide_Quit_Menu();
+            g_theControlBar->Toggle_Purchase_Science();
+        } else if (id == buttonLargeID) {
+            g_theControlBar->Toggle_Control_Bar_Stage();
+        } else if (id == buttonOptions) {
+            Toggle_Quit_Menu();
+        } else if (id == buttonIdleWorker) {
+            Hide_Quit_Menu();
+            g_theInGameUI->Select_Next_Idle_Worker();
+        } else {
+            g_theControlBar->Process_Context_Sensitive_Button_Click(window2, static_cast<GadgetGameMessage>(message));
+        }
+
+        return MSG_HANDLED;
+    }
+
+    if (message == GWM_CREATE) {
+        buttonCommunicator = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:PopupCommunicator");
+        return MSG_HANDLED;
+    }
+
+    if (message <= GGM_CLOSE) {
+        return MSG_IGNORED;
+    }
+
+    g_theControlBar->Process_Context_Sensitive_Button_Transition(window2, static_cast<GadgetGameMessage>(message));
+    return MSG_HANDLED;
 }
 
-void Show_Control_Bar(bool show)
+void Toggle_Control_Bar(bool immediate)
```

Comment:
Perhaps move this func impl closer to Hide_Control_Bar, Show_Control_Bar

Reply:
fixed.